### PR TITLE
Validate our configuration by using a schema.

### LINKED ISSUE #1317: Enable validation of our configuration via yaml schema
Inspired by [this blog post](https://blog.picnic.nl/how-to-use-yaml-schema-to-validate-your-yaml-files-c82c049c2097) I added a first "very WIP" branch [here](https://github.com/troessner/reek/pull/1316) which I would like to pursue after we have done #1078 - thoughts?

### PR DESCRIPTION
**Pull request by troessner**
Fixes #1317

### CODE REVIEW

File: conf/schema.yml [hunk 1]
```diff
@@ -0,0 +1,196 @@
...
+        mapping:
+          <<: *detector_base
+      UtilityFunction:
+        type: map
+        mapping:
+          <<: *detector_base
+          public_methods_only:
+            type: bool
+
+  "directories":
```
Comment (troessner) [inline conf/schema.yml]: Actually kwalify seems to be pretty awesome. Those half a dozen lines below validate all directory directives ❤️

File: lib/reek/configuration/converter.rb [hunk 2]
```diff
@@ -26,7 +26,6 @@ def strings_to_regexes(configuration)
           return configuration unless configuration[AppConfiguration::DETECTORS_KEY]
           detector_configuration = configuration[AppConfiguration::DETECTORS_KEY]
           detector_configuration.keys.
-            select { |key| smell_type?(key) }.
```
Comment (troessner) [inline lib/reek/configuration/converter.rb]: This check can go now with schema validation 🎉

File: lib/reek/configuration/default_directive.rb [hunk 3]
```diff
@@ -21,7 +21,6 @@ module DefaultDirective
       # @return [self]
       def add(detectors_configuration)
         detectors_configuration.each do |name, configuration|
-          next unless smell_type?(name)
```
Comment (troessner) [inline lib/reek/configuration/default_directive.rb]: And this one as well 🎈

File: lib/reek/configuration/schema_validator.rb [hunk 4]
```diff
@@ -0,0 +1,32 @@
+# frozen_string_literal: true
+
+require 'yaml'
+require_relative '../cli/silencer'
+Reek::CLI::Silencer.silently { require 'kwalify' }
```
Comment (troessner) [inline lib/reek/configuration/schema_validator.rb]: Yes, I need to fully qualify this here, CLI::Silencer won't do.

**Comment by troessner** [comment]
Ready for review!

**Comment by mvz** [comment]
I wasn't happy with all the silencers there, so I did a little proof of concept for only suppressing the warnings and that seems to work well: f956d8913aca0902bf6a433253235f5c73b03613. Wdyt?

File: conf/schema.yml [hunk 5]
```diff
@@ -0,0 +1,210 @@
+---
```
Comment (mvz) [inline conf/schema.yml]: How about keeping this in `lib/Reek/Configuration`, similar to where we store the HTML template?

File: features/configuration_files/directory_specific_directives.feature [hunk 6]
```diff
@@ -200,10 +200,7 @@ Feature: Directory directives
     Then the exit status indicates an error
     And stderr reports:
     """
-    You are trying to configure smell type IteratorsNested but we can't find one with that name.
-    Please make sure you spelled it right. (See 'docs/defaults.reek' in the Reek
-    repository for a list of all available smell types.)
-
+    Error: Bad configuration file: key 'IteratorsNested:' is undefined.
```
Comment (mvz) [inline features/configuration_files/directory_specific_directives.feature]: This is somewhat less user-friendly. Perhaps we should enhance `ConfigFileError` to provide some guidance.

File: features/configuration_files/schema_validation.feature [hunk 7]
```diff
@@ -0,0 +1,247 @@
+Feature: Validate schema
```
Comment (mvz) [inline features/configuration_files/schema_validation.feature]: Can we move some of what is tested here to specs and just have one or two scenarios here to check the default configuration and the failure case?

File: features/step_definitions/sample_file_steps.rb [hunk 8]
```diff
@@ -37,6 +37,11 @@
   write_file(filename, CONFIGURATION_DIR.join(filename).read)
 end
 
+Given(/^our default configuration file '(.+)'$/) do |filename|
+  default_configuration = File.read SAMPLES_DIR.join('..').join('docs').join('defaults.reek')
```
Comment (mvz) [inline features/step_definitions/sample_file_steps.rb]: `'defaults.reek'` -> `filename`?

File: lib/reek/configuration/configuration_file_finder.rb [hunk 9]
```diff
@@ -63,6 +64,7 @@ def load_from_file(path)
           unless configuration.is_a? Hash
             raise Errors::ConfigFileError, "Invalid configuration file \"#{path}\" -- Not a hash"
           end
+          SchemaValidator.new(configuration).validate
```
Comment (mvz) [inline lib/reek/configuration/configuration_file_finder.rb]: Do we still need the Hash check above?

**Comment by troessner** [comment]
>> I wasn't happy with all the silencers there, so I did a little proof of concept for only suppressing the warnings and that seems to work well: f956d89. Wdyt?

Very nice. Separate PR after this one?

**Comment by mvz** [comment]
Almost there :tada:

File: features/step_definitions/sample_file_steps.rb (hunk 8 above)
Comment (troessner) [reply features/step_definitions/sample_file_steps.rb]: Fixed

File: features/configuration_files/directory_specific_directives.feature (hunk 6 above)
Comment (mvz) [reply features/configuration_files/directory_specific_directives.feature]: Yes, looks good.

**Comment by troessner** [comment]
@mvz addressed all your feedback! Merge? :)

**Comment by mvz** [comment]
There are some failures ...

**Comment by troessner** [comment]
And green it is ;)

**Comment by mvz** [comment]
Awesome!

**Pull request merged**